pkg/ccl/backupccl/tenant_backup_nemesis_test.go

```diff
@@ -0,0 +1,505 @@
+			t.Logf("backup-nemesis: incremental backup finished")
+		}
+		return nil
+	})
+
+	require.NoError(t, g.Wait())
+	restoreQuery := fmt.Sprintf("RESTORE TENANT 10 FROM LATEST IN '%s' AS OF SYSTEM TIME %s WITH tenant = '11'", backupLoc, aost)
+	t.Logf("backup-nemesis: restoring tenant 10 into 11: %s", restoreQuery)
+	hostSQLDB.Exec(t, restoreQuery)
+
+	// Validation
+	//
+	// We check bank.bank which has had the workload running against it
+	// and any table from a completed nemesis.
+	tenant11, err := tc.Servers[0].StartTenant(ctx, base.TestTenantArgs{
+		TenantID:            roachpb.MustMakeTenantID(11),
+		DisableCreateTenant: true,
+	})
+	require.NoError(t, err)
+
+	tenant11Conn, err := serverutils.OpenDBConnE(
+		tenant11.SQLAddr(), "bank", false, tenant11.Stopper())
+	require.NoError(t, err)
+
+	tenant10SQLDB := sqlutils.MakeSQLRunner(tenant10Conn)
+	tenant11SQLDB := sqlutils.MakeSQLRunner(tenant11Conn)
+
+	validateTableAcrossTenantRestore(t, tenant10SQLDB, tenant11SQLDB, "bank.bank", aost)
+	for _, name := range tablesToCheck {
+		validateTableAcrossTenantRestore(t, tenant10SQLDB, tenant11SQLDB, fmt.Sprintf(`bank."%s"`, name), aost)
+	}
+}
+
+func validateTableAcrossTenantRestore(
+	t *testing.T,
+	oldTenant *sqlutils.SQLRunner,
+	newTenant *sqlutils.SQLRunner,
+	tableName string,
+	endTime string,
+) {
+	t.Logf("backup-nemesis: checking table %s at %s", tableName, endTime)
+	assertEqualQueriesAtTime(t, oldTenant, newTenant,
+		fmt.Sprintf("SELECT count(*) FROM %s", tableName), endTime)
+	assertEqualQueriesAtTime(t, oldTenant, newTenant,
+		fmt.Sprintf("SELECT * FROM [SHOW EXPERIMENTAL_FINGERPRINTS FROM TABLE %s]", tableName), endTime)
+}
+
+func assertEqualQueriesAtTime(
+	t *testing.T,
+	oldTenant *sqlutils.SQLRunner,
+	newTenant *sqlutils.SQLRunner,
+	queryBase string,
+	endTime string,
+) {
+	t.Helper()
+	assertEqualQueries(t, oldTenant, newTenant, fmt.Sprintf("%s AS OF SYSTEM TIME %s", queryBase, endTime))
+}
+
+func assertEqualQueries(
+	t *testing.T, db1 *sqlutils.SQLRunner, db2 *sqlutils.SQLRunner, query string, args ...interface{},
+) {
+	t.Helper()
+	out1 := db1.QueryStr(t, query, args...)
+	out2 := db2.QueryStr(t, query, args...)
+	require.Equal(t, out1, out2)
+}
+
+type randomBackupNemesis struct {
+	t  *testing.T
+	db *gosql.DB
+
+	grp    ctxgroup.Group
+	cancel context.CancelFunc
+	rng    *rand.Rand
+
+	nemeses []nemesis
+
+	mu struct {
+		syncutil.Mutex
+		oneTimeListeners []chan nemesisNotification
+		tablesToCheck    []string
+	}
+}
+
+type nemesis struct {
+	name string
+	impl func(context.Context, *randomBackupNemesis, *gosql.DB) error
+}
+
+type nemesisNotification struct {
+	name string
+	done chan struct{}
+}
+
+func newBackupNemesis(t *testing.T, rng *rand.Rand, db *gosql.DB) *randomBackupNemesis {
+	return &randomBackupNemesis{
+		t:   t,
+		rng: rng,
+		db:  db,
+
+		nemeses: []nemesis{
+			{name: "CANCELED IMPORT INTO",
+				impl: func(ctx context.Context, n *randomBackupNemesis, db *gosql.DB) error {
```

Comment:
the number of incremental backups the import job spanned had some affect on the corruption cases we saw. I wonder if we could randomly vary the speed of the import job, or any of the jobs below.

Reply:
This seems like it would be a fun thing to add.  The other thing I was thinking of adding to the "listener" stuff was something where the operations could let you know when an "interesting" point in the operations was occurring and then allow you to block the operation, take a backup, and continue.

Reply:
Well, this would require a good deal more infrastructure such as that which is already calling confusion. So I'll leave this to whoever is motivated to touch this test next.